ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -13,161 +13,304 @@
+public class Log implements Write {
+  private final String dataDir;
   private final long capacityInBytes;
-  private static final String Log_File_Name = "log_current";
-  private Logger logger = LoggerFactory.getLogger(getClass());
   private final StoreMetrics metrics;
+  private final AtomicLong remainingUnallocatedCapacity;
+  private final ConcurrentSkipListMap<String, LogSegment> segmentsByName =
+      new ConcurrentSkipListMap<>(LogSegmentNameHelper.COMPARATOR);
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  private LogSegment activeSegment;
 
-  public Log(String dataDir, long capacityInBytes, StoreMetrics metrics)
+  /**
+   * Create a Log instance
+   * @param dataDir the directory where the segments of the log need to be created.
+   * @param totalCapacityInBytes the total capacity of this log.
+   * @param segmentCapacityInBytes the capacity of a single segment in the log.
+   * @param metrics the {@link StoreMetrics} instance to use.
+   * @throws IOException if there is any I/O error loading the segment files.
+   * @throws IllegalArgumentException if {@code totalCapacityInBytes} or {@code segmentCapacityInBytes} <= 0 or if
+   * {@code totalCapacityInBytes} > {@code segmentCapacityInBytes} and {@code totalCapacityInBytes} is not a perfect
+   * multiple of {@code segmentCapacityInBytes}.
+   */
+  public Log(String dataDir, long totalCapacityInBytes, long segmentCapacityInBytes, StoreMetrics metrics)
       throws IOException {
-    file = new File(dataDir, Log_File_Name);
-    if (!file.exists()) {
-      // if the file does not exist, preallocate it
-      Utils.preAllocateFileIfNeeded(file, capacityInBytes);
+    if (totalCapacityInBytes <= 0 || segmentCapacityInBytes <= 0) {
+      throw new IllegalArgumentException(
+          "One of totalCapacityInBytes [" + totalCapacityInBytes + "] or " + "segmentCapacityInBytes ["
+              + segmentCapacityInBytes + "] is <=0");
+    }
+    long segmentCapacity = Math.min(totalCapacityInBytes, segmentCapacityInBytes);
+    // all segments should be the same size.
+    if (totalCapacityInBytes % segmentCapacity != 0) {
+      throw new IllegalArgumentException(
+          "Capacity of log [" + totalCapacityInBytes + "] should be a perfect multiple of segment capacity ["
+              + segmentCapacityInBytes + "]");
     }
-    this.capacityInBytes = capacityInBytes;
-    fileChannel = Utils.openChannel(file, true);
-    logger.trace("Log : {} file size on start {} ", dataDir, fileChannel.size());
-    // A log's write offset will always be set to the start of the log.
-    // External components is responsible for setting it the right value
-    currentWriteOffset = new AtomicLong(0);
+    this.dataDir = dataDir;
+    this.capacityInBytes = totalCapacityInBytes;
     this.metrics = metrics;
+    remainingUnallocatedCapacity = new AtomicLong(totalCapacityInBytes);
+    long numSegments = (totalCapacityInBytes - 1) / segmentCapacityInBytes + 1;
```

Comment:
since you ensure `totalCapacityInBytes` is a multiple of `segmentCapacityInBytes`, why not keep this simple and write `numSegments = totalCapacityInBytes/segmentCapacityInBytes`?